tests/workers/test_base_worker.py

```diff
@@ -1719,3 +1723,92 @@ async def test_get_flow_run_logger(
             "work_pool_name": "test-work-pool",
             "work_pool_id": str(work_pool.id),
         }
+
+
+class TestInfrastructureIntegration:
+    @pytest.fixture
+    def mock_infrastructure_run(self, monkeypatch) -> Generator[MagicMock, None, None]:
+        """
+        Mocks all subtype implementations of `Infrastructure.run`.
+
+        Yields a mock that is called with `self.dict()` when `run`
+        is awaited. The mock provides a few utilities for testing
+        error handling.
+
+        `pre_start_side_effect` and `post_start_side_effect` may be
+        set to callables to perform actions before or after the
+        task is reported as started.
+
+        `mark_as_started` may be set to `False` to prevent marking the
+        task as started.
+        """
+        mock = MagicMock()
+        mock.pre_start_side_effect = lambda: None
+        mock.post_start_side_effect = lambda: None
+        mock.mark_as_started = True
+        mock.result_status_code = 0
+        mock.result_identifier = "id-1234"
+
+        async def mock_run(self, task_status=None):
+            # Record the call immediately
+            result = mock(self.dict())
+            result.status_code = mock.result_status_code
+            result.identifier = mock.result_identifier
+
+            # Perform side-effects for testing error handling
+
+            pre = mock.pre_start_side_effect()
+            if inspect.iscoroutine(pre):
+                await pre
+
+            if mock.mark_as_started:
+                task_status.started(result.identifier)
+
+            post = mock.post_start_side_effect()
+            if inspect.iscoroutine(post):
+                await post
+
+            return result
+
+        # Patch all infrastructure types
+        types = get_registry_for_type(Block)
+        for t in types.values():
+            if not issubclass(t, Infrastructure):
+                continue
+            monkeypatch.setattr(t, "run", mock_run)
```

Comment:
It looks like you're mocking out `run` on infrastructure blocks instead of workers. I think you could simplify the setup by creating a test `BaseWorker` subclass like `WorkerTestImpl` but with mock run like the one you have defined above.

Reply:
Thanks for the suggestion @desertaxle ! Indeed, the test could looks much simpler like this. Please let me know your thoughts! 

PS - I'm still a bit puzzled as to how the previous test failed on CI, but passes perfectly on my local. Any clues on why this might happen?